hazelcast-client/src/main/java/com/hazelcast/client/proxy/ClientPNCounterProxy.java

```diff
@@ -0,0 +1,451 @@
+        final VectorClock timestamps = new VectorClock();
+        for (Entry<String, Long> replicaTimestamp : replicaLogicalTimestamps) {
+            timestamps.setReplicaTimestamp(replicaTimestamp.getKey(), replicaTimestamp.getValue());
+        }
+        return timestamps;
+    }
+
+    /**
+     * Adds the {@code delta} and returns the value of the counter before the
+     * update if {@code getBeforeUpdate} is {@code true} or the value after
+     * the update if it is {@code false}.
+     * It will invoke client messages recursively on viable replica addresses
+     * until successful or the list of viable replicas is exhausted.
+     * Replicas with addresses contained in the {@code excludedAddresses} are
+     * skipped. If there are no viable replicas, this method will throw the
+     * {@code lastException} if not {@code null} or an {@link MutationDisallowedException}
+     * if the {@code lastException} is {@code null}.
+     *
+     * @param delta             the delta to add to the counter value, can be negative
+     * @param getBeforeUpdate   {@code true} if the operation should return the
+     *                          counter value before the addition, {@code false}
+     *                          if it should return the value after the addition
+     * @param excludedAddresses the addresses to exclude when choosing a replica
+     *                          address, may be {@code null}
+     * @param lastException     the exception thrown from the last invocation of
+     *                          the {@code request} on a replica, may be {@code null}
+     * @return the result of the request invocation on a replica
+     * @throws MutationDisallowedException if there are no replicas and the
+     *                                     {@code lastException} is false
+     */
+    private ClientMessage invokeAddInternal(long delta, boolean getBeforeUpdate,
+                                            List<Address> excludedAddresses,
+                                            HazelcastException lastException,
+                                            Address target) {
+        if (target == null) {
+            throw lastException != null
+                    ? lastException
+                    : new MutationDisallowedException(
+                    "Cannot invoke operations on a CRDT because the cluster does not contain any data members");
+        }
+        try {
+            final ClientMessage request = PNCounterAddCodec.encodeRequest(
+                    name, delta, getBeforeUpdate, observedClock.entrySet(), target);
+            return invokeOnAddress(request, target);
+        } catch (ConsistencyLostException e) {
+            logger.fine("Unable to provide session guarantees when sending operations to " + target
+                    + ", choosing different target");
+            if (excludedAddresses == EMPTY_ADDRESS_LIST) {
```

Comment:
It may be a good idea to shrink these 4 repetitive lines below into a single utility method, this will raise readability, but this's just a matter of taste.

Reply:
I was thinking about that, seemed like a small gain for the extra method but I'll try again

Reply:
Extracted the method, as well as for the `ClientPNCounterProxy` and the `PNCounterProxy`